From 887bacf7402b4c92cd6bd2e74e34e62dda4e591a Mon Sep 17 00:00:00 2001
From: Valentin Caron <valentin.caron@foss.st.com>
Date: Fri, 13 Dec 2024 11:35:53 +0100
Subject: [PATCH 1302/1695] drivers: clk: stm32mp15: add is_enabled ops

This ops return the real status of the gate: enabled or not.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 510e4f241bf887d734a0727b221e093f71c90978 drivers: clk: stm32mp15: add is_enabled ops]
Signed-off-by: Valentin Caron <valentin.caron@foss.st.com>
Change-Id: I75584366c109de2a38a5107ea972732f7f5ddbc7
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/424685
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/438469
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Tested-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/drivers/clk/clk-stm32mp15.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/core/drivers/clk/clk-stm32mp15.c b/core/drivers/clk/clk-stm32mp15.c
index 94e343695..490466b3b 100644
--- a/core/drivers/clk/clk-stm32mp15.c
+++ b/core/drivers/clk/clk-stm32mp15.c
@@ -1969,12 +1969,23 @@ static void clk_op_disable(struct clk *clk)
 }
 DECLARE_KEEP_PAGER(clk_op_disable);
 
+static bool clk_op_is_enabled(struct clk *clk)
+{
+	if (clk_is_gate(clk))
+		return __clk_is_enabled(clk_to_gate_ref(clk));
+
+	return refcount_val(&clk->enabled_count) != 0;
+
+}
+DECLARE_KEEP_PAGER(clk_op_is_enabled);
+
 /* This variable is weak to break its dependency chain when linked as unpaged */
 const struct clk_ops stm32mp1_clk_ops
 __weak __relrodata_unpaged("stm32mp1_clk_ops") = {
 	.enable = clk_op_enable,
 	.disable = clk_op_disable,
 	.get_rate = clk_op_compute_rate,
+	.is_enabled = clk_op_is_enabled,
 };
 
 static TEE_Result register_stm32mp1_clocks(void)
-- 
2.47.2

